lighthouse-core/scripts/i18n/cull-obsolete-lhl-messages.js

```diff
@@ -0,0 +1,152 @@
+/**
+ * @license Copyright 2019 Google Inc. All Rights Reserved.
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
+ */
+'use strict';
+
+const fs = require('fs');
+const path = require('path');
+const glob = require('glob');
+const MessageParser = require('intl-messageformat-parser').default;
+
+const {collectAllCustomElementsFromICU} = require('../../lib/i18n/i18n.js');
+
+/** @typedef {Record<string, {message: string}>} LhlMessages */
+
+/**
+ * Returns whether the string `lhlMessage` has ICU arguments matching the
+ * already extracted `goldenArgumentIds`. Assumes `goldenArgumentIds` is sorted.
+ * @param {Array<string>} goldenArgumentIds
+ * @param {string} lhlMessage
+ * @return {boolean}
+ */
+function equalArguments(goldenArgumentIds, lhlMessage) {
+  const parsedMessage = MessageParser.parse(lhlMessage);
+  const lhlArgumentElements = collectAllCustomElementsFromICU(parsedMessage.elements);
+  const lhlArgumentIds = [...lhlArgumentElements.keys()];
+
+  if (goldenArgumentIds.length !== lhlArgumentIds.length) return false;
+
+  lhlArgumentIds.sort();
+  for (let i = 0; i < goldenArgumentIds.length; i++) {
+    if (goldenArgumentIds[i] !== lhlArgumentIds[i]) {
+      return false;
+    }
+  }
+
+  return true;
+}
+
+/**
+ * Logs a message as removed if it hasn't been logged before.
+ * @param {Set<string>} alreadyLoggedCulls
+ * @param {string} messageId
+ * @param {string} reason
+ */
+function logRemoval(alreadyLoggedCulls, messageId, reason) {
+  if (alreadyLoggedCulls.has(messageId)) return;
+
+  // eslint-disable-next-line no-console
+  console.log(`Removing message\n\t'${messageId}'\nfrom translations: ${reason}.`);
+  alreadyLoggedCulls.add(messageId);
+}
+
+/**
+ * Returns a copy of `localeLhl` with only messages matching those from the golden locale.
+ * `goldenLocaleArgumentIds` values are assumed to be sorted.
+ * @param {Record<string, Array<string>>} goldenLocaleArgumentIds
+ * @param {LhlMessages} localeLhl
+ * @param {Set<string>} alreadyLoggedCulls Set of culls that have been logged and shouldn't be logged again.
+ * @return {LhlMessages}
+ */
+function cullLocale(goldenLocaleArgumentIds, localeLhl, alreadyLoggedCulls) {
+  /** @type {LhlMessages} */
+  const remainingMessages = {};
+
+  for (const [messageId, {message}] of Object.entries(localeLhl)) {
+    const goldenArgumentIds = goldenLocaleArgumentIds[messageId];
+    if (!goldenArgumentIds) {
+      logRemoval(alreadyLoggedCulls, messageId, 'it is no longer found in Lighthouse');
```

Comment:
As a moderate bonus, old UIStrings that have been removed from Lighthouse are removed from all locales, though this will eventually happen regardless when new tc results come in.